enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,141 @@
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+- Use the self-hosted stack also for UPI deployments.
+
+
+### Non-Goals
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/on-prem) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run the self-hosted stack components (Keepalived, Coredns-MDNS, HAProxy, MDNS-publisher) as a K8S applications.
+
+### Kubelet and Operator circular dependency: 
+In order to run the operator's component, Kubelet in each node should first join the cluster by communicating with the control plane, however the components provided by the operator are responsible for enabling that.
+As a result, the below circular dependency should be addressed:
+* Kubelet can't talk to the control plane until the operator has started.
+* Operator can't start until Kubelet can talk to the control plane.
+
+To break the circular dependency, the operator should start running before bootstrap is completed in order to assure that bootstrap owns the API-VIP. In addition, a new /etc/hosts entry should be added to the nodes using MCO ignition.  This /etc/hosts entry is expected to resolve api-int hostname to the API-VIP address. 
+
+### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+
+```yaml
+apiVersion: clusterhosted.operator.io/v1alpha1
+kind: ClusterHostedConfig
+metadata:
+  name: clusterhostedconfig-sample
+spec:
+  dns: Enable
+  loadbalancer: Enable
+status:
+  apiserverinternalipowner: master-0
+  ingressipowner: worker-1
+```
+
+This manifest creates DNS and loadbalancer self-hosted stack, from the status section we can see that master-0 node currently owns api-int VIP and worker-1 node owns ingress VIP. Additional spec and status fields may be introduced based on requirements.
+
+### User Stories
+
+#### As an admin I want to disable the self-hosted on-prem LB and DNS stack.
+
+#### As an admin I want to use the self-hosted on-prem LB and DNS stack also in UPI deployments.
+
+#### As an admin I want to retrieve the self-hosted stack status through K8S API.
+
+
+
+### Upgrade / Downgrade Strategy
+
+Prior to this proposal, the self-hosted stack implementation was running through the MCO.
+The implemntation of the self-hostd stack using the new operator might lead to cases during upgrade/downgrade in which two parallel components are running the self-hosted stack. 
+
```

Comment:
Again referring to the CBO upgrade/downgrade strategy in https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md may be helpful, since it's a similar situation (moving control from MAO to a new SLO)